src/gsIO/gsFileManager.cpp

```diff
@@ -70,96 +57,157 @@ bool gsFileManager::fileExists(const std::string& name)
-        || util::starts_with(fn,".\\")
-        || util::starts_with(fn,"../")
-        || util::starts_with(fn,"..\\");
+#if defined _WIN32 || defined __CYGWIN__
+    static std::string ps(":*?\"<>|"); // ':' allowed at fn[1]
 #else
-    return util::starts_with(fn,"./")
-        || util::starts_with(fn,"../");
+    // allowed under Unix: \?%*:|"<>
+    static std::string ps("");
 #endif
+    return ps;
 }
 
-void _replace_slash_by_basckslash(std::string& str)
+namespace {
+inline bool _contains( const std::string& haystack, char needle )
 {
-    for ( std::string::iterator it=str.begin(); it!=str.end(); it++ )
-        if ( *it=='/' ) *it = '\\';
+    return haystack.find(needle) != std::string::npos;
 }
 
-bool gsFileManager::addSearchPaths(const std::string& paths)
+bool _isValidPathOrName(const std::string& fn)
 {
-    return gsFileManagerDataSingleton().addSearchPaths(paths);
+    bool invalid = false;
+    for (size_t i = 0; i < fn.length(); ++i)
+    {
+        invalid = invalid | _contains(gsFileManager::getInvalidCharacters(), fn[i]);
+    }
+    return !invalid;
 }
 
-bool gsFileManager::setSearchPaths(const std::string& paths)
+// sets last character to the native path seperator
+// special case "" gets "./"
+void _makePath(std::string& final_result)
+{
+	if (final_result.length() == 0)
+		final_result.push_back('.');
+	if (final_result[final_result.length() - 1] != gsFileManager::getNativePathSeparator())
+		final_result.push_back(gsFileManager::getNativePathSeparator());
+}
+
+} // anonymous namespace
+
+bool gsFileManager::isFullyQualified(const std::string& fn)
+{
+    bool valid = false;
+#if defined _WIN32 || defined __CYGWIN__
+    // case "c:\\abc"
+    if ( fn.size() > 2 && isalpha(fn[0]) && fn[1] == ':')
+    {
+        for (int i = 0; i < getValidPathSeparators().length(); ++i)
+        {
+            valid = valid || (fn[2] == getValidPathSeparators()[i]);
+        }
+        valid = valid && _isValidPathOrName(fn.substr(3, fn.length()-3));
+    }
+    else // case "\\abc", same as Unix
+    {
+#endif
+    for (size_t i = 0; i < getValidPathSeparators().length(); ++i)
+    {
+        valid = valid || (fn[0] == getValidPathSeparators()[i]);
+    }
+    valid = valid && _isValidPathOrName(fn);
+#if defined _WIN32 || defined __CYGWIN__
+    }
+#endif
+    return valid;
+}
+
+bool gsFileManager::isExplicitlyRelative(const std::string& fn)
 {
-    gsFileManagerDataSingleton().clear();
-    return gsFileManagerDataSingleton().addSearchPaths(paths);
+    bool valid = false;
+    for (size_t i = 0; i < getValidPathSeparators().length(); ++i)
+    {
+        valid = valid || (fn[0] == '.' && fn[1] == getValidPathSeparators()[i]) ||
+         (fn[0] == '.' && fn[1] == '.' && fn[2] == getValidPathSeparators()[i]);
+    }
+    valid = valid && _isValidPathOrName(fn);
+    return valid;
 }
 
-bool gsFileManagerData::addSearchPaths(const std::string& paths)
+namespace {
+
+inline void _repacle_with_native_seperator(std::string& str)
```

Comment:
rename as:
_replacle_with_native_seperator

Reply:
`_replace_with_native_separator` - days of typos